helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java

```diff
@@ -1085,6 +1085,139 @@ public ClusterManagementMode getClusterManagementMode(String clusterName) {
         : new ClusterManagementMode(status.getManagementMode(), status.getManagementModeStatus());
   }
 
+  private enum SetPartitionToErrorFailureReason {
+    INSTANCE_NOT_ALIVE("%s is not alive in cluster %s"),
+    INSTANCE_NON_EXISTENT("%s does not exist in cluster %s"),
+    RESOURCE_NON_EXISTENT("resource %s is not added to cluster %s"),
+    PARTITION_NON_EXISTENT("not all %s exist in cluster %s"),
+    PARTITION_ALREADY_IN_ERROR("%s is NOT found in cluster %s or is already in ERROR state"),
+    STATE_MODEL_NON_EXISTENT("%s is NOT found in cluster %s");
+
+    private String message;
+
+    SetPartitionToErrorFailureReason(String message) {
+      this.message = message;
+    }
+
+    public String getMessage(String resourceName, List<String> partitionNames, String instanceName,
+        String errorStateEntity, String clusterName) {
+      return String.format("Can't set to Error State for %s.%s on %s, because " + message, resourceName,
+          partitionNames, instanceName, errorStateEntity, clusterName);
+    }
+  }
+
+  @Override
+  public void setPartitionToError(String clusterName, String instanceName, String resourceName,
+      List<String> partitionNames) {
+    logger.info("Set partitions {} for resource {} on instance {} in cluster {} to ERROR state.",
+        partitionNames == null ? "NULL" : HelixUtil.serializeByComma(partitionNames), resourceName, instanceName,
+        clusterName);
+    HelixDataAccessor accessor = new ZKHelixDataAccessor(clusterName, new ZkBaseDataAccessor<ZNRecord>(_zkClient));
+    PropertyKey.Builder keyBuilder = accessor.keyBuilder();
+
+    // check the instance is alive
+    LiveInstance liveInstance = accessor.getProperty(keyBuilder.liveInstance(instanceName));
+    if (liveInstance == null) {
+      // check if the instance exists in the cluster
+      String instanceConfigPath = PropertyPathBuilder.instanceConfig(clusterName, instanceName);
+      throw new HelixException(String.format(
+          (_zkClient.exists(instanceConfigPath) ? SetPartitionToErrorFailureReason.INSTANCE_NOT_ALIVE
+              : SetPartitionToErrorFailureReason.INSTANCE_NON_EXISTENT).getMessage(resourceName, partitionNames,
+              instanceName, instanceName, clusterName)));
+    }
+
+    // check resource exists in ideal state
+    IdealState idealState = accessor.getProperty(keyBuilder.idealStates(resourceName));
+      if (idealState == null) {
+      throw new HelixException(String.format(SetPartitionToErrorFailureReason.RESOURCE_NON_EXISTENT
+          .getMessage(resourceName, partitionNames, instanceName, resourceName, clusterName)));
+    }
+
+    // check partition exists in resource
+    Set<String> partitionsToBeSetToError = new HashSet<String>(partitionNames);
+    Set<String> partitions =
```

Comment:
from what i understand, the RebalanceMode used is independent of the state model and the state transitions. Since we are sending a state transition message directly to the participant in these two cases - reset and setPartitionsToError, we can allow it for all modes.